File: reverse-vowels-of-string/reverse-vowel.js
/*https://leetcode.com/problems/reverse-vowels-of-a-string/*/

/**
 * @param {string} s
 * @return {string}
 */
 var reverseVowels = function(s) {
    const sList = [...s];
    const vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
    let vowelList = [];  
    for (let i = 0; i < s.length; i++) {
        if (vowels.includes(s[i])) {
            vowelList.push(s[i])
        }
    }

    for (let k = 0; k < s.length; k++) {
        if (vowels.includes(s[k])) {
            sList[k] = vowelList.pop();
        }
    };
    return sList.join('');
};

/**
 * @param {string} s
 * @return {string}
 */
 var reverseVowels = function(s) {
    const sList = [...s];
    const vowels = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
    let vowelList = [];
    for (let i = 0; i < s.length; i++) {
        if (vowels.includes(s[i])) {
            vowelList.push(s[i])
        }
    }
    const reversedList = vowelList.reverse();
    let index = 0;
    for (let k = 0; k < s.length; k++) {
        if (vowels.includes(s[k])) {
            sList[k] = reversedList[index];
            index += 1;
        }
    };
    return sList.join('');
};

File: reverse-vowels-of-string/reverse-vowel.py
#https://leetcode.com/problems/reverse-vowels-of-a-string/

class Solution:
    def reverseVowels(self, s: str) -> str:
        vowel_list = ['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
        list_of_s = list(s)
        left = 0
        right = len(s) - 1
        while left < right:
            while left < right and list_of_s[left] not in vowel_list:
                left += 1
            while right > left and list_of_s[right] not in vowel_list:
                right -= 1
            left_vowel = list_of_s[left]
            list_of_s[left] = list_of_s[right]
            list_of_s[right] = left_vowel
            left += 1
            right -= 1
        return ''.join(list_of_s)